File: src/ui/MissionResultViewModel.js
import { MISSION_OUTCOME_COMMANDS as COMMANDS } from '../missions/MissionOutcomeService.js';
import { clone, deepFreeze } from '../missions/ContractUtils.js';

export const MISSION_RESULT_KINDS = Object.freeze({
  SUCCESS: 'SUCCESS',
  PARTIAL_SUCCESS: 'PARTIAL_SUCCESS',
  FAILURE: 'FAILURE',
  ABANDONED: 'ABANDONED',
  ARRESTED: 'ARRESTED',
  VEHICLE_LOSS: 'VEHICLE_LOSS'
});

export const RESULT_SECTION_IDS = Object.freeze({
  REWARD: 'REWARD',
  CITY: 'CITY',
  FACTION: 'FACTION',
  PROGRESSION: 'PROGRESSION'
});

const OUTCOME_ALIASES = Object.freeze({
  SUCCESS: MISSION_RESULT_KINDS.SUCCESS,
  COMPLETE: MISSION_RESULT_KINDS.SUCCESS,
  COMPLETED: MISSION_RESULT_KINDS.SUCCESS,
  PARTIAL: MISSION_RESULT_KINDS.PARTIAL_SUCCESS,
  PARTIAL_SUCCESS: MISSION_RESULT_KINDS.PARTIAL_SUCCESS,
  ABANDONED: MISSION_RESULT_KINDS.ABANDONED,
  ABANDONMENT: MISSION_RESULT_KINDS.ABANDONED,
  CANCELLED: MISSION_RESULT_KINDS.ABANDONED,
  CANCELED: MISSION_RESULT_KINDS.ABANDONED,
  ARREST: MISSION_RESULT_KINDS.ARRESTED,
  ARRESTED: MISSION_RESULT_KINDS.ARRESTED,
  CAPTURED: MISSION_RESULT_KINDS.ARRESTED,
  VEHICLE_LOSS: MISSION_RESULT_KINDS.VEHICLE_LOSS,
  VEHICLE_LOST: MISSION_RESULT_KINDS.VEHICLE_LOSS,
  VEHICLE_DESTROYED: MISSION_RESULT_KINDS.VEHICLE_LOSS,
  FAILURE: MISSION_RESULT_KINDS.FAILURE,
  FAILED: MISSION_RESULT_KINDS.FAILURE
});

const KIND_PRESENTATION = Object.freeze({
  [MISSION_RESULT_KINDS.SUCCESS]: Object.freeze({ label: 'Success', tone: 'success' }),
  [MISSION_RESULT_KINDS.PARTIAL_SUCCESS]: Object.freeze({ label: 'Partial success', tone: 'partial' }),
  [MISSION_RESULT_KINDS.FAILURE]: Object.freeze({ label: 'Mission failed', tone: 'failure' }),
  [MISSION_RESULT_KINDS.ABANDONED]: Object.freeze({ label: 'Mission abandoned', tone: 'abandoned' }),
  [MISSION_RESULT_KINDS.ARRESTED]: Object.freeze({ label: 'Operator arrested', tone: 'arrested' }),
  [MISSION_RESULT_KINDS.VEHICLE_LOSS]: Object.freeze({ label: 'Vehicle lost', tone: 'vehicle-loss' })
});

const REASON_EXPLANATIONS = Object.freeze({
  timeout: 'The mission timer expired before the final objective was completed.',
  cancelled: 'You abandoned the mission before its final objective was completed.',
  canceled: 'You abandoned the mission before its final objective was completed.',
  abandoned: 'You abandoned the mission before its final objective was completed.',
  released: 'Control of the required mission vehicle was released during the run.',
  vehicle_lost: 'The required mission vehicle was lost, destroyed, or exchanged during the run.',
  vehicle_destroyed: 'The required mission vehicle was destroyed during the run.',
  arrest: 'Enforcement captured the operator before the mission could be completed.',
  arrested: 'Enforcement captured the operator before the mission could be completed.',
  captured: 'Enforcement captured the operator before the mission could be completed.',
  race_lost: 'A rival reached the finish before the player.'
});

const EFFECT_LABELS = Object.freeze({
  [COMMANDS.CAPITAL_ADJUSTED]: 'Capital',
  [COMMANDS.BUILDING_STATE_SET]: 'Building state',
  [COMMANDS.INFRASTRUCTURE_STATE_SET]: 'Infrastructure',
  [COMMANDS.INCIDENT_RECORDED]: 'Incident',
  [COMMANDS.INCIDENT_RESOLVED]: 'Incident resolved',
  [COMMANDS.REPAIR_SET]: 'Repair status',
  [COMMANDS.SERVICE_OUTAGE_SET]: 'Service coverage',
  [COMMANDS.TRAFFIC_SET]: 'Traffic conditions',
  [COMMANDS.FACTION_REPUTATION_ADJUSTED]: 'Reputation',
  [COMMANDS.PROGRESSION_SET]: 'Career tier',
  [COMMANDS.UNLOCK_SET]: 'Capability',
  [COMMANDS.NEWS_PUBLISHED]: 'City news',
  [COMMANDS.FOLLOW_UP_MISSION_SET]: 'Follow-up mission',
  [COMMANDS.AUTHORED_FLAG_SET]: 'City state'
});

const SECTION_META = Object.freeze({
  [RESULT_SECTION_IDS.REWARD]: Object.freeze({
    title: 'Reward & performance',
    empty: 'No Capital or performance reward was recorded.'
  }),
  [RESULT_SECTION_IDS.CITY]: Object.freeze({
    title: 'City consequences',
    empty: 'No persistent city condition changed.'
  }),
  [RESULT_SECTION_IDS.FACTION]: Object.freeze({
    title: 'Faction consequences',
    empty: 'No faction reputation changed.'
  }),
  [RESULT_SECTION_IDS.PROGRESSION]: Object.freeze({
    title: 'Progression & unlocks',
    empty: 'No progression, capability, or follow-up unlocked.'
  })
});

const PROGRESSION_TYPES = new Set([
  COMMANDS.PROGRESSION_SET,
  COMMANDS.UNLOCK_SET,
  COMMANDS.FOLLOW_UP_MISSION_SET
]);

function normalizedToken(value) {
  return String(value ?? '').trim().toUpperCase().replace(/[\s-]+/g, '_');
}

function humanizeId(value) {
  const text = String(value ?? '').trim();
  if (!text) return 'Unknown';
  const words = text.replace(/[._-]+/g, ' ');
  const normalized = words === words.toUpperCase() ? words.toLowerCase() : words;
  return normalized
    .replace(/\b\w/g, letter => letter.toUpperCase());
}

function finite(value) {
  return typeof value === 'number' && Number.isFinite(value);
}

function signedNumber(value) {
  return `${value > 0 ? '+' : ''}${value.toLocaleString('en-US')}`;
}

function currency(value) {
  const absolute = Math.abs(value).toLocaleString('en-US');
  return `${value < 0 ? '-' : ''}$${absolute}`;
}

function percentage(value) {
  return `${Math.round(value * 100)}%`;
}

function primaryRecordValue(record) {
  if (record == null || typeof record !== 'object') return record;
  if (typeof record.headline === 'string') return record.headline;
  if (typeof record.status === 'string') return humanizeId(record.status);
  if (typeof record.state === 'string') return humanizeId(record.state);
  if (typeof record.access === 'string') return `${humanizeId(record.access)} access`;
  if (typeof record.active === 'boolean') return record.active ? 'Active' : 'Resolved';
  if (typeof record.unlocked === 'boolean') return record.unlocked ? 'Unlocked' : 'Locked';
  if (finite(record.coverageMultiplier)) return `${percentage(record.coverageMultiplier)} coverage`;
  if (finite(record.densityMultiplier)) return `${record.densityMultiplier.toFixed(2)}× traffic`;
  if (finite(record.condition)) return `${percentage(record.condition)} condition`;
  if (Object.hasOwn(record, 'value')) return String(record.value);
  return null;
}

function formatChange(effect) {
  const before = effect.before;
  const after = effect.after;
  if (effect.type === COMMANDS.CAPITAL_ADJUSTED && finite(before) && finite(after)) {
    const delta = after - before;
    return `${currency(before)} → ${currency(after)} (${delta >= 0 ? '+' : '-'}${currency(Math.abs(delta))})`;
  }
  if (effect.type === COMMANDS.FACTION_REPUTATION_ADJUSTED && finite(before) && finite(after)) {
    return `${before.toLocaleString('en-US')} → ${after.toLocaleString('en-US')} (${signedNumber(after - before)})`;
  }
  if (typeof after === 'boolean') return after ? 'Unlocked' : 'Locked';
  if (finite(before) && finite(after)) return `${before.toLocaleString('en-US')} → ${after.toLocaleString('en-US')} (${signedNumber(after - before)})`;
  const beforeValue = primaryRecordValue(before);
  const afterValue = primaryRecordValue(after);
  if (beforeValue != null && afterValue != null && beforeValue !== afterValue) return `${beforeValue} → ${afterValue}`;
  if (afterValue != null) return String(afterValue);
  if (beforeValue != null && after == null) return `${beforeValue} → Resolved`;
  return before == null ? 'Recorded' : 'Updated';
}

function classifyKind(resolution = {}, source = {}) {
  const reasonToken = normalizedToken(resolution.reason ?? source.reason);
  if (reasonToken === 'CANCELLED' || reasonToken === 'CANCELED' || reasonToken === 'ABANDONED') {
    return MISSION_RESULT_KINDS.ABANDONED;
  }
  if (reasonToken === 'ARREST' || reasonToken === 'ARRESTED' || reasonToken === 'CAPTURED') {
    return MISSION_RESULT_KINDS.ARRESTED;
  }
  if (['VEHICLE_LOSS', 'VEHICLE_LOST', 'VEHICLE_DESTROYED'].includes(reasonToken)) {
    return MISSION_RESULT_KINDS.VEHICLE_LOSS;
  }
  return OUTCOME_ALIASES[normalizedToken(resolution.outcome ?? source.outcome)] || MISSION_RESULT_KINDS.FAILURE;
}

function sectionForEffect(type) {
  if (type === COMMANDS.CAPITAL_ADJUSTED) return RESULT_SECTION_IDS.REWARD;
  if (type === COMMANDS.FACTION_REPUTATION_ADJUSTED) return RESULT_SECTION_IDS.FACTION;
  if (PROGRESSION_TYPES.has(type)) return RESULT_SECTION_IDS.PROGRESSION;
  return RESULT_SECTION_IDS.CITY;
}

function createEffectItem(effect, labels = {}) {
  const subjectLabel = labels[effect.subjectId] || humanizeId(effect.subjectId);
  return {
    id: `${effect.type}:${effect.subjectId}`,
    type: effect.type,
    label: effect.type === COMMANDS.CAPITAL_ADJUSTED
      ? EFFECT_LABELS[effect.type]
      : `${EFFECT_LABELS[effect.type] || humanizeId(effect.type)} · ${subjectLabel}`,
    value: formatChange(effect),
    explanation: effect.explanation || 'The committed mission outcome changed this value.'
  };
}

function performanceItems(resolution = {}) {
  const items = [];
  if (finite(resolution.satisfaction)) {
    items.push({
      id: 'performance:satisfaction',
      type: 'SATISFACTION',
      label: 'Passenger satisfaction',
      value: `${Math.round(resolution.satisfaction)}%`,
      explanation: 'Satisfaction reflects time used and traffic encountered during the run.'
    });
  }
  if (finite(resolution.damage)) {
    items.push({
      id: 'performance:damage',
      type: 'DAMAGE',
      label: 'Damage',
      value: resolution.damage <= 1 ? percentage(resolution.damage) : resolution.damage.toLocaleString('en-US'),
      explanation: 'Damage recorded when the mission resolved.'
    });
  }
  if (finite(resolution.heat)) {
    items.push({
      id: 'performance:heat',
      type: 'HEAT',
      label: 'Heat',
      value: resolution.heat.toLocaleString('en-US'),
      explanation: 'Enforcement Heat recorded when the mission resolved.'
    });
  }
  return items;
}

function buildWhy(resolution, explanation, weather) {
  const rawReason = resolution.reason ?? explanation.source?.reason;
  const key = String(rawReason ?? '').trim().toLowerCase();
  const primary = REASON_EXPLANATIONS[key]
    || (typeof rawReason === 'string' && rawReason.trim() ? humanizeId(rawReason) : explanation.description);
  const details = [primary];
  if (weather?.disposition === 'ADAPTED' && weather.reason && weather.reason !== primary) details.push(weather.reason);
  return [...new Set(details.filter(Boolean))];
}

function buildNextAction(kind, retry) {
  if (retry?.allowed) {
    const remaining = finite(retry.attemptsRemaining) ? ` ${retry.attemptsRemaining} attempt${retry.attemptsRemaining === 1 ? '' : 's'} remain.` : '';
    return {
      title: 'Retry or return to the city',
      description: `${retry.reason || 'Restart from the mission approach.'}${remaining}`,
      canRetry: true,
      retryLabel: retry.checkpoint ? 'Retry from checkpoint' : 'Retry mission',
      continueLabel: 'Return to Management'
    };
  }
  if (kind === MISSION_RESULT_KINDS.SUCCESS || kind === MISSION_RESULT_KINDS.PARTIAL_SUCCESS) {
    return {
      title: 'Review the changed city',
      description: 'Return to Management to inspect the committed consequences and choose the next priority.',
      canRetry: false,
      retryLabel: 'Retry mission',
      continueLabel: 'Return to Management'
    };
  }
  return {
    title: 'Recover in Management',
    description: retry?.reason || 'Return to Management. The committed result remains in the outcome log.',
    canRetry: false,
    retryLabel: 'Retry mission',
    continueLabel: 'Return to Management'
  };
}

function explanationFromReceipt(receipt) {
  return {
    transactionId: receipt.transactionId,
    source: clone(receipt.source),
    title: receipt.summary?.title || 'Recorded outcome',
    description: receipt.summary?.description || receipt.source?.reason || 'A persistent outcome was committed.',
    effects: clone(receipt.effects || [])
  };
}

export function buildMissionResultViewModel({
  lifecycleSnapshot = null,
  explanation,
  mission = null,
  retryDecision = null,
  labels = {},
  sequence = null
} = {}) {
  if (!explanation || typeof explanation !== 'object') throw new TypeError('A result explanation is required');
  const run = lifecycleSnapshot?.run || null;
  const resolution = run?.resolution || {};
  const kind = classifyKind(resolution, explanation.source);
  const presentation = KIND_PRESENTATION[kind];
  const sectionItems = new Map(Object.values(RESULT_SECTION_IDS).map(id => [id, []]));
  for (const effect of explanation.effects || []) {
    sectionItems.get(sectionForEffect(effect.type)).push(createEffectItem(effect, labels));
  }
  sectionItems.get(RESULT_SECTION_IDS.REWARD).push(...performanceItems(resolution));
  const sections = Object.values(RESULT_SECTION_IDS).map(id => ({
    id,
    ...SECTION_META[id],
    items: sectionItems.get(id)
  }));
  const nextAction = buildNextAction(kind, retryDecision);
  const missionTitle = mission?.title || explanation.title || humanizeId(explanation.source?.contentId);
  const why = buildWhy(resolution, explanation, run?.weather);
  const changeCount = sections.reduce((total, section) => total + section.items.length, 0);
  const announcement = [
    `${presentation.label}: ${missionTitle}.`,
    explanation.description,
    `${changeCount} recorded change${changeCount === 1 ? '' : 's'}.`,
    nextAction.description
  ].filter(Boolean).join(' ');

  return deepFreeze({
    transactionId: explanation.transactionId,
    sequence,
    kind,
    outcomeLabel: presentation.label,
    tone: presentation.tone,
    missionTitle,
    title: explanation.title || missionTitle,
    description: explanation.description || 'The mission outcome was committed.',
    why,
    attempt: run?.attempt ?? null,
    sections,
    nextAction,
    announcement
  });
}

export function buildMissionResultHistory({ receipts = [], lifecycleSnapshot = null, missions = [], labels = {} } = {}) {
  const missionMap = new Map(missions.map(mission => [mission.id, mission]));
  const currentTransactionId = lifecycleSnapshot?.run?.transactionId;
  const entries = [...receipts]
    .filter(receipt => receipt.source?.kind === 'MISSION')
    .sort((left, right) => (right.sequence || 0) - (left.sequence || 0))
    .map(receipt => buildMissionResultViewModel({
      lifecycleSnapshot: receipt.transactionId === currentTransactionId ? lifecycleSnapshot : null,
      explanation: explanationFromReceipt(receipt),
      mission: missionMap.get(receipt.source?.contentId) || null,
      labels,
      sequence: receipt.sequence ?? null
    }));
  return deepFreeze(entries);
}

export function classifyMissionResult({ resolution = {}, source = {} } = {}) {
  return classifyKind(resolution, source);
}

File: godot/MetroPulse.Domain/Missions/MissionResultViewModel.cs
using System.Globalization;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace MetroPulse.Domain.Missions;

/// <summary>Receipt-time result and persistent history projection; never reads mutable current city state.</summary>
public static partial class MissionResultViewModel
{
    private static readonly string[] SectionOrder =
        [MissionResultSectionIds.Reward, MissionResultSectionIds.City, MissionResultSectionIds.Faction, MissionResultSectionIds.Progression];

    private static readonly IReadOnlyDictionary<string, (string Label, string Tone)> Presentations =
        new Dictionary<string, (string, string)>(StringComparer.Ordinal)
        {
            [MissionResultKinds.Success] = ("Success", "success"),
            [MissionResultKinds.PartialSuccess] = ("Partial success", "partial"),
            [MissionResultKinds.Failure] = ("Mission failed", "failure"),
            [MissionResultKinds.Abandoned] = ("Mission abandoned", "abandoned"),
            [MissionResultKinds.Arrested] = ("Operator arrested", "arrested"),
            [MissionResultKinds.VehicleLoss] = ("Vehicle lost", "vehicle-loss"),
        };

    private static readonly IReadOnlyDictionary<string, (string Title, string Empty)> SectionMetadata =
        new Dictionary<string, (string, string)>(StringComparer.Ordinal)
        {
            [MissionResultSectionIds.Reward] = ("Reward & performance", "No Capital or performance reward was recorded."),
            [MissionResultSectionIds.City] = ("City consequences", "No persistent city condition changed."),
            [MissionResultSectionIds.Faction] = ("Faction consequences", "No faction reputation changed."),
            [MissionResultSectionIds.Progression] = ("Progression & unlocks", "No progression, capability, or follow-up unlocked."),
        };

    private static readonly IReadOnlyDictionary<string, string> EffectLabels =
        new Dictionary<string, string>(StringComparer.Ordinal)
        {
            [OutcomeCommandTypes.CapitalAdjusted] = "Capital",
            [OutcomeCommandTypes.BuildingStateSet] = "Building state",
            [OutcomeCommandTypes.InfrastructureStateSet] = "Infrastructure",
            [OutcomeCommandTypes.IncidentRecorded] = "Incident",
            [OutcomeCommandTypes.IncidentResolved] = "Incident resolved",
            [OutcomeCommandTypes.RepairSet] = "Repair status",
            [OutcomeCommandTypes.ServiceOutageSet] = "Service coverage",
            [OutcomeCommandTypes.TrafficSet] = "Traffic conditions",
            [OutcomeCommandTypes.FactionReputationAdjusted] = "Reputation",
            [OutcomeCommandTypes.ProgressionSet] = "Career tier",
            [OutcomeCommandTypes.UnlockSet] = "Capability",
            [OutcomeCommandTypes.NewsPublished] = "City news",
            [OutcomeCommandTypes.FollowUpMissionSet] = "Follow-up mission",
            [OutcomeCommandTypes.AuthoredFlagSet] = "City state",
        };

    private static readonly IReadOnlyDictionary<string, string> ReasonExplanations =
        new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
        {
            ["timeout"] = "The mission timer expired before the final objective was completed.",
            ["cancelled"] = "You abandoned the mission before its final objective was completed.",
            ["canceled"] = "You abandoned the mission before its final objective was completed.",
            ["abandoned"] = "You abandoned the mission before its final objective was completed.",
            ["released"] = "Control of the required mission vehicle was released during the run.",
            ["vehicle_lost"] = "The required mission vehicle was lost, destroyed, or exchanged during the run.",
            ["vehicle_destroyed"] = "The required mission vehicle was destroyed during the run.",
            ["arrest"] = "Enforcement captured the operator before the mission could be completed.",
            ["arrested"] = "Enforcement captured the operator before the mission could be completed.",
            ["captured"] = "Enforcement captured the operator before the mission could be completed.",
            ["race_lost"] = "A rival reached the finish before the player.",
        };

    public static MissionResultView Build(
        MissionOutcomeExplanation explanation,
        MissionLifecycleState? lifecycle = null,
        MissionDefinition? mission = null,
        MissionRetryDecision? retry = null,
        long? sequence = null,
        IReadOnlyDictionary<string, string>? labels = null)
    {
        ArgumentNullException.ThrowIfNull(explanation);
        if (lifecycle is not null)
        {
            if (lifecycle.Phase != MissionPhases.Result || lifecycle.Run?.TransactionId != explanation.TransactionId
                || lifecycle.Run.Receipt?.TransactionId != explanation.TransactionId)
                throw new InvalidDataException("A current result view requires the matching committed RESULT receipt.");
        }
        MissionResolution? resolution = lifecycle?.Run?.Resolution;
        string kind = Classify(resolution, explanation.Source);
        (string outcomeLabel, string tone) = Presentations[kind];
        var items = SectionOrder.ToDictionary(section => section, _ => new List<MissionResultEffectItem>(), StringComparer.Ordinal);
        foreach (OutcomeEffect effect in explanation.Effects)
        {
            items[SectionFor(effect.Type)].Add(CreateEffectItem(effect, labels));
        }
        AddPerformance(items[MissionResultSectionIds.Reward], resolution);
        MissionResultSection[] sections = SectionOrder.Select(id => new MissionResultSection(
            id,
            SectionMetadata[id].Title,
            SectionMetadata[id].Empty,
            Array.AsReadOnly(items[id].ToArray()))).ToArray();
        MissionResultNextAction next = BuildNextAction(kind, retry);
        string missionTitle = mission?.Title ?? explanation.Title ?? Humanize(explanation.Source.ContentId);
        string[] why = BuildWhy(resolution, explanation, lifecycle?.Run?.Weather);
        int changeCount = sections.Sum(section => section.Items.Count);
        string announcement = string.Join(' ', new[]
        {
            $"{outcomeLabel}: {missionTitle}.",
            explanation.Description,
            $"{changeCount} recorded change{(changeCount == 1 ? string.Empty : "s")}.",
            next.Description,
        }.Where(value => !string.IsNullOrWhiteSpace(value)));
        return new MissionResultView(
            explanation.TransactionId,
            sequence,
            kind,
            outcomeLabel,
            tone,
            missionTitle,
            explanation.Title ?? missionTitle,
            explanation.Description ?? "The mission outcome was committed.",
            Array.AsReadOnly(why),
            lifecycle?.Run?.Attempt,
            Array.AsReadOnly(sections),
            next,
            announcement);
    }

    public static IReadOnlyList<MissionResultView> BuildHistory(
        IEnumerable<MissionOutcomeReceipt> receipts,
        MissionRegistry registry,
        MissionLifecycleState? currentLifecycle = null,
        IReadOnlyDictionary<string, string>? labels = null)
    {
        ArgumentNullException.ThrowIfNull(receipts);
        ArgumentNullException.ThrowIfNull(registry);
        MissionResultView[] history = receipts
            .Where(receipt => receipt.Source.Kind == OutcomeSourceKinds.Mission)
            .OrderByDescending(receipt => receipt.Sequence)
            .Select(receipt => Build(
                new MissionOutcomeExplanation(
                    receipt.TransactionId,
                    receipt.Source,
                    receipt.Summary.Title,
                    receipt.Summary.Description,
                    receipt.Effects),
                currentLifecycle?.Run?.TransactionId == receipt.TransactionId ? currentLifecycle : null,
                registry.Get(receipt.Source.ContentId),
                sequence: receipt.Sequence,
                labels: labels))
            .ToArray();
        return Array.AsReadOnly(history);
    }

    public static string Classify(MissionResolution? resolution, OutcomeSource source)
    {
        ArgumentNullException.ThrowIfNull(source);
        string reason = Token(resolution?.Reason ?? source.Reason);
        if (reason is "CANCELLED" or "CANCELED" or "ABANDONED") return MissionResultKinds.Abandoned;
        if (reason is "ARREST" or "ARRESTED" or "CAPTURED") return MissionResultKinds.Arrested;
        if (reason is "VEHICLE_LOSS" or "VEHICLE_LOST" or "VEHICLE_DESTROYED") return MissionResultKinds.VehicleLoss;
        return Token(resolution?.Outcome ?? source.Outcome) switch
        {
            "SUCCESS" or "COMPLETE" or "COMPLETED" => MissionResultKinds.Success,
            "PARTIAL" or "PARTIAL_SUCCESS" => MissionResultKinds.PartialSuccess,
            "ABANDONED" or "ABANDONMENT" or "CANCELLED" or "CANCELED" => MissionResultKinds.Abandoned,
            "ARREST" or "ARRESTED" or "CAPTURED" => MissionResultKinds.Arrested,
            "VEHICLE_LOSS" or "VEHICLE_LOST" or "VEHICLE_DESTROYED" => MissionResultKinds.VehicleLoss,
            _ => MissionResultKinds.Failure,
        };
    }

    private static void AddPerformance(List<MissionResultEffectItem> items, MissionResolution? resolution)
    {
        if (resolution?.Satisfaction is { } satisfaction)
            items.Add(new("performance:satisfaction", "SATISFACTION", "Passenger satisfaction", $"{Math.Round(satisfaction):N0}%", "Satisfaction reflects time used and traffic encountered during the run."));
        if (resolution?.Damage is { } damage)
            items.Add(new("performance:damage", "DAMAGE", "Damage", damage <= 1 ? $"{Math.Round(damage * 100):N0}%" : FormatNumber(damage), "Damage recorded when the mission resolved."));
        if (resolution?.Heat is { } heat)
            items.Add(new("performance:heat", "HEAT", "Heat", FormatNumber(heat), "Enforcement Heat recorded when the mission resolved."));
    }

    private static MissionResultEffectItem CreateEffectItem(OutcomeEffect effect, IReadOnlyDictionary<string, string>? labels)
    {
        string subject = labels?.GetValueOrDefault(effect.SubjectId) ?? Humanize(effect.SubjectId);
        string baseLabel = EffectLabels.GetValueOrDefault(effect.Type) ?? Humanize(effect.Type);
        string label = effect.Type == OutcomeCommandTypes.CapitalAdjusted ? baseLabel : $"{baseLabel} · {subject}";
        return new MissionResultEffectItem(
            $"{effect.Type}:{effect.SubjectId}",
            effect.Type,
            label,
            FormatChange(effect),
            string.IsNullOrWhiteSpace(effect.Explanation) ? "The committed mission outcome changed this value." : effect.Explanation);
    }

    private static string SectionFor(string type) => type switch
    {
        OutcomeCommandTypes.CapitalAdjusted => MissionResultSectionIds.Reward,
        OutcomeCommandTypes.FactionReputationAdjusted => MissionResultSectionIds.Faction,
        OutcomeCommandTypes.ProgressionSet or OutcomeCommandTypes.UnlockSet or OutcomeCommandTypes.FollowUpMissionSet => MissionResultSectionIds.Progression,
        _ => MissionResultSectionIds.City,
    };

    private static string[] BuildWhy(MissionResolution? resolution, MissionOutcomeExplanation explanation, MissionWeatherDecision? weather)
    {
        string? raw = resolution?.Reason ?? explanation.Source.Reason;
        string primary = raw is not null && ReasonExplanations.TryGetValue(raw, out string? known)
            ? known
            : !string.IsNullOrWhiteSpace(raw) ? Humanize(raw) : explanation.Description;
        return new[] { primary, weather?.Disposition == MissionWeatherDispositions.Adapted ? weather.Reason : null }
            .Where(value => !string.IsNullOrWhiteSpace(value))
            .Distinct(StringComparer.Ordinal)
            .Select(value => value!)
            .ToArray();
    }

    private static MissionResultNextAction BuildNextAction(string kind, MissionRetryDecision? retry)
    {
        if (retry?.Allowed == true)
        {
            string remaining = retry.AttemptsRemaining is { } count
                ? $" {count} attempt{(count == 1 ? string.Empty : "s")} remain."
                : string.Empty;
            return new MissionResultNextAction(
                "Retry or return to the city",
                $"{retry.Reason}{remaining}",
                true,
                retry.Checkpoint is null ? "Retry mission" : "Retry from checkpoint",
                "Return to Management");
        }
        if (kind is MissionResultKinds.Success or MissionResultKinds.PartialSuccess)
            return new MissionResultNextAction(
                "Review the changed city",
                "Return to Management to inspect the committed consequences and choose the next priority.",
                false,
                "Retry mission",
                "Return to Management");
        return new MissionResultNextAction(
            "Recover in Management",
            retry?.Reason ?? "Return to Management. The committed result remains in the outcome log.",
            false,
            "Retry mission",
            "Return to Management");
    }

    private static string FormatChange(OutcomeEffect effect)
    {
        if (effect.Type == OutcomeCommandTypes.CapitalAdjusted && Numbers(effect, out double before, out double after))
        {
            double delta = after - before;
            return $"{Currency(before)} → {Currency(after)} ({(delta >= 0 ? "+" : "-")}{Currency(Math.Abs(delta))})";
        }
        if (effect.Type == OutcomeCommandTypes.FactionReputationAdjusted && Numbers(effect, out before, out after))
            return $"{FormatNumber(before)} → {FormatNumber(after)} ({(after - before > 0 ? "+" : string.Empty)}{FormatNumber(after - before)})";
        if (effect.After is { ValueKind: JsonValueKind.True or JsonValueKind.False } boolean) return boolean.GetBoolean() ? "Unlocked" : "Locked";
        if (Numbers(effect, out before, out after))
            return $"{FormatNumber(before)} → {FormatNumber(after)} ({(after - before > 0 ? "+" : string.Empty)}{FormatNumber(after - before)})";
        string? beforeValue = PrimaryValue(effect.Before);
        string? afterValue = PrimaryValue(effect.After);
        if (beforeValue is not null && afterValue is not null && beforeValue != afterValue) return $"{beforeValue} → {afterValue}";
        if (afterValue is not null) return afterValue;
        if (beforeValue is not null && effect.After is null) return $"{beforeValue} → Resolved";
        return effect.Before is null ? "Recorded" : "Updated";
    }

    private static string? PrimaryValue(JsonElement? element)
    {
        if (element is null || element.Value.ValueKind is JsonValueKind.Null or JsonValueKind.Undefined) return null;
        JsonElement value = element.Value;
        if (value.ValueKind == JsonValueKind.String) return value.GetString();
        if (value.ValueKind is JsonValueKind.True or JsonValueKind.False) return value.GetBoolean() ? "Active" : "Resolved";
        if (value.ValueKind == JsonValueKind.Number && value.TryGetDouble(out double number)) return FormatNumber(number);
        if (value.ValueKind != JsonValueKind.Object) return value.ToString();
        foreach (string key in new[] { "headline", "status", "state", "access" })
        {
            if (value.TryGetProperty(key, out JsonElement property) && property.ValueKind == JsonValueKind.String)
            {
                string result = Humanize(property.GetString());
                return key == "access" ? $"{result} access" : result;
            }
        }
        if (value.TryGetProperty("active", out JsonElement active) && active.ValueKind is JsonValueKind.True or JsonValueKind.False)
            return active.GetBoolean() ? "Active" : "Resolved";
        if (value.TryGetProperty("unlocked", out JsonElement unlocked) && unlocked.ValueKind is JsonValueKind.True or JsonValueKind.False)
            return unlocked.GetBoolean() ? "Unlocked" : "Locked";
        if (value.TryGetProperty("coverageMultiplier", out JsonElement coverage) && coverage.TryGetDouble(out number))
            return $"{Math.Round(number * 100):N0}% coverage";
        if (value.TryGetProperty("densityMultiplier", out JsonElement density) && density.TryGetDouble(out number))
            return $"{number:F2}× traffic";
        if (value.TryGetProperty("condition", out JsonElement condition) && condition.TryGetDouble(out number))
            return $"{Math.Round(number * 100):N0}% condition";
        if (value.TryGetProperty("value", out JsonElement authoredValue)) return authoredValue.ToString();
        return null;
    }

    private static bool Numbers(OutcomeEffect effect, out double before, out double after)
    {
        before = 0;
        after = 0;
        return effect.Before is { ValueKind: JsonValueKind.Number } beforeElement && beforeElement.TryGetDouble(out before)
            && effect.After is { ValueKind: JsonValueKind.Number } afterElement && afterElement.TryGetDouble(out after);
    }

    private static string Currency(double value) => $"{(value < 0 ? "-" : string.Empty)}${Math.Abs(value).ToString("N0", CultureInfo.GetCultureInfo("en-US"))}";

    private static string FormatNumber(double value) => value.ToString(value % 1 == 0 ? "N0" : "0.###", CultureInfo.GetCultureInfo("en-US"));

    private static string Token(string? value) => Regex().Replace((value ?? string.Empty).Trim().ToUpperInvariant(), "_");

    private static string Humanize(string? value)
    {
        string text = (value ?? string.Empty).Trim();
        if (text.Length == 0) return "Unknown";
        string words = Regex().Replace(text, " ").ToLowerInvariant();
        return CultureInfo.GetCultureInfo("en-US").TextInfo.ToTitleCase(words);
    }

    [GeneratedRegex("[\\s._-]+")]
    private static partial Regex Regex();
}
